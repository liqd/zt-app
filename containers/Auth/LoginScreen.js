import React, { useEffect, useState } from 'react'
import { Alert, Linking, ScrollView, View } from 'react-native'
import { Formik } from 'formik'
import * as yup from 'yup'

import API, { baseUrl } from '../../BaseApi'
import { ButtonSubmit } from '../../components/ButtonSubmit'
import { TextInputFormField } from '../../components/formFields'
import { LinkTextSourceSans } from '../../components/LinkTextSourceSans'
import { TextSourceSans } from '../../components/TextSourceSans'

import {useAuthorization} from './AuthProvider.js'
import { styles } from './LoginScreen.styles'

export const LoginScreen = () => {
  const {signIn} = useAuthorization()
  const [error, setError] = useState()
  const registerUrl = '/accounts/signup/?next=/'
  const passwordResetUrl = '/accounts/password/reset/'

  useEffect(() => {
    if (error) {
      Alert.alert('An error occured', error, [{ text: 'Ok' }])
    }
  }, [error])

  const handleLogin = (values) => {
    API.postLogin(values).then((response) => {
      if (response.statusCode!==200) {
        if (response.data.non_field_errors) {
          setError(response.data.non_field_errors[0])
        } else if (response.data.username) {
          setError('Username or E-mail address: ' + response.data.username[0])
        } else if (response.data.password) {
          setError('Password: ' + response.data.password[0])
        } else {
          setError('Something went wrong, please try again.')
        }
      } else {
        signIn(response.data.token)
      }
    })
  }

  const loginValidationSchema = yup.object().shape({
    username: yup
      .string()
      .required('Please enter your e-mail address or username.'),
    password: yup
      .string()
      .required('Please enter your password.'),
  })

  return (
    <Formik
      validationSchema={loginValidationSchema}
      initialValues={{ username: '', password: '' }}
      onSubmit={values => handleLogin(values)}
    >
      {({
        handleChange,
        handleBlur,
        handleSubmit,
        values,
        errors,
        touched,
        isValid
      }) => (
        <View style={styles.flexContainer}>
          <View style={styles.container}>
            <ScrollView>
              <TextSourceSans style={styles.title}>
                Login
              </TextSourceSans>
              <TextSourceSans style={styles.registerText}>
                If you have not created an account yet, then please <LinkTextSourceSans
                  onPress={() => Linking.openURL(baseUrl + registerUrl)}
                >register</LinkTextSourceSans> first.
              </TextSourceSans>
              <TextInputFormField
                field='Username or E-mail address:'
                name='username'
                value={values.username}
                placeholder=''
                returnKeyType='next'
                returnKeyLabel='next'
                onChangeText={handleChange('username')}
                onBlur={handleBlur('username')}
                error={errors.username}
                touched={touched.username}
                testID='username-input'
              />
              <TextInputFormField
                field='Password:'
                name='password'
                value={values.password}
                placeholder=''
                returnKeyType='next'
                returnKeyLabel='next'
                onChangeText={handleChange('password')}
                onBlur={handleBlur('password')}
                error={errors.password}
                touched={touched.password}
                secureTextEntry
                testID='password-input'
              />
              <LinkTextSourceSans style={styles.forgotPassword}
                onPress={() => Linking.openURL(baseUrl + passwordResetUrl)}
              >Forgot Password?</LinkTextSourceSans>
            </ScrollView>
          </View>
          <ButtonSubmit
            title='Login'
            onPress={handleSubmit}
            disabled={!isValid}
          >
          </ButtonSubmit>
        </View>
      )}
    </Formik>
  )
}
