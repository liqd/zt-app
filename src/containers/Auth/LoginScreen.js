import React, { useEffect, useState } from 'react'
import { Trans, useTranslation } from 'react-i18next'
import { Alert, Linking, View } from 'react-native'
import { Formik } from 'formik'
import * as yup from 'yup'

import API, { baseUrl } from '../../BaseApi'
import { TextInputFormField } from '../../components/formFields'
import { KeyboardScrollView } from '../../components/KeyboardScrollView'
import { LinkTextSourceSans } from '../../components/LinkTextSourceSans'
import { StatusBarStyled } from '../../components/StatusBarStyled'
import { TextSourceSans } from '../../components/TextSourceSans'

import {useAuthorization} from './AuthProvider.js'
import { styles } from './LoginScreen.styles'

export const LoginScreen = () => {
  const { t } = useTranslation()
  const {signIn} = useAuthorization()
  const [error, setError] = useState()
  const [submitPending, setSubmitPending] = useState()
  const registerUrl = '/accounts/signup/?next=/'
  const passwordResetUrl = '/accounts/password/reset/'

  useEffect(() => {
    if (error) {
      Alert.alert(t('An error occured'), error, [{ text: 'Ok' }])
    }
  }, [error])

  const handleLogin = (values) => {
    setSubmitPending(true)
    return API.postLogin(values)
      .then((response) => {
        const {statusCode, data} = response
        if (statusCode === 200) {
          signIn(data.token)
        } else if (statusCode === 400) {
          let errorMessage
          if (data.non_field_errors) {
            errorMessage = data.non_field_errors[0]
          } else if (data.username) {
            errorMessage = t('Username or E-mail address: ') + data.username[0]
          } else if (data.password) {
            errorMessage = t('Password: ') + data.password[0]
          } else {
            errorMessage = t('Something went wrong, please try again.')
          }
          return Promise.reject(errorMessage)
        }
      })
      .catch(error => {
        setError(error)
        setSubmitPending(false)
      })
  }

  const loginValidationSchema = yup.object().shape({
    username: yup
      .string()
      .required(t('Please enter your e-mail address or username.')),
    password: yup
      .string()
      .required(t('Please enter your password.')),
  })

  return (
    <Formik
      validationSchema={loginValidationSchema}
      initialValues={{ username: '', password: '' }}
      onSubmit={values => handleLogin(values)}
      validateOnMount={true}
    >
      {({
        handleChange,
        handleBlur,
        handleSubmit,
        values,
        errors,
        touched,
        isValid
      }) => (
        <KeyboardScrollView
          handleSubmit={handleSubmit}
          isValid={isValid && !submitPending}
          buttonText='Login'
        >
          <StatusBarStyled />
          <View style={styles.container}>
            <TextSourceSans style={styles.title}>
              {t('Login')}
            </TextSourceSans>
            <TextSourceSans style={styles.registerText}>
              <Trans t={t}>
                If you have not created an account yet, then please <LinkTextSourceSans
                  onPress={() => Linking.openURL(baseUrl + registerUrl)}
                >register</LinkTextSourceSans> first.
              </Trans>
            </TextSourceSans>
            <TextInputFormField
              field={t('Username or E-mail address:')}
              name='username'
              value={values.username}
              placeholder=''
              returnKeyType='next'
              returnKeyLabel='next'
              onChangeText={handleChange('username')}
              onBlur={handleBlur('username')}
              error={errors.username}
              touched={touched.username}
              testID='username-input'
            />
            <TextInputFormField
              field={t('Password:')}
              name='password'
              value={values.password}
              placeholder=''
              returnKeyType='next'
              returnKeyLabel='next'
              onChangeText={handleChange('password')}
              onBlur={handleBlur('password')}
              error={errors.password}
              touched={touched.password}
              secureTextEntry
              testID='password-input'
            />
            <LinkTextSourceSans style={styles.forgotPassword}
              onPress={() => Linking.openURL(baseUrl + passwordResetUrl)}
            >{t('Forgot Password?')}</LinkTextSourceSans>
          </View>
        </KeyboardScrollView>
      )}
    </Formik>
  )
}
