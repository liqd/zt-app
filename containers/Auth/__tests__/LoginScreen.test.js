import React from 'react';
import '@testing-library/jest-native/extend-expect';
import { render, waitFor, screen, fireEvent } from '@testing-library/react-native';
import { LoginScreen } from '../LoginScreen';
import API from '../../../BaseApi';

jest.mock('react-native/Libraries/Animated/NativeAnimatedHelper');

test('Test LoginScreen Snapshot', () => {
  const { toJSON } = render(<LoginScreen />);
  expect(toJSON()).toMatchSnapshot();
});

test('Test Login, no credentials', async () => {
  render(<LoginScreen />);
  const submitButton = screen.getByRole('button');
  fireEvent(submitButton, 'press');
  const validationErrors = await screen.findAllByText(/Please enter your password/);
  expect(validationErrors).toBeTruthy();
});

test('Test Login, correct credentials', async () => {
  API.postLogin = jest.fn(() => Promise.resolve({
    statusCode: 200,
    data: { token: 'test-token'}
  }));

  render(<LoginScreen />);
  const submitButton = screen.getByRole('button');
  const nameInput = screen.getByTestId('username-input');
  const pwInput = screen.getByTestId('password-input');
  const username = 'testuser';
  const password = 'testPassword';

  fireEvent(nameInput, 'changeText', username);
  expect(nameInput).toHaveProp('value', username);
  fireEvent(pwInput, 'changeText', password);
  expect(pwInput).toHaveProp('value', password);
  fireEvent.press(submitButton);
  await waitFor(() => expect(API.postLogin).toHaveBeenCalledWith({password: password, username: username}));
});
