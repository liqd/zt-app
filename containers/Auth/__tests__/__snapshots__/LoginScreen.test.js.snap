// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test LoginScreen Snapshot 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#fff",
      "flex": 1,
      "justifyContent": "space-between",
    }
  }
>
  <View
    style={
      Object {
        "paddingHorizontal": 15,
      }
    }
  >
    <RCTScrollView>
      <View>
        <Text
          style={
            Array [
              Object {
                "fontFamily": "SourceSansPro_400Regular",
              },
              Object {
                "fontSize": 28,
                "fontWeight": "800",
                "marginBottom": 15,
                "marginTop": 80,
              },
            ]
          }
        >
          Login
        </Text>
        <Text
          style={
            Array [
              Object {
                "fontFamily": "SourceSansPro_400Regular",
              },
              Object {
                "fontSize": 16,
                "marginBottom": 30,
              },
            ]
          }
        >
          If you have not created an account yet, then please 
          <Text
            onPress={[Function]}
            style={
              Array [
                Object {
                  "fontFamily": "SourceSansPro_400Regular",
                },
                Object {
                  "textDecorationLine": "underline",
                },
              ]
            }
          >
            register
          </Text>
           first.
        </Text>
        <View>
          <Text
            style={
              Array [
                Object {
                  "fontFamily": "SourceSansPro_400Regular",
                },
                Object {
                  "fontSize": 15,
                  "lineHeight": 19,
                  "paddingBottom": 5,
                  "paddingTop": 10,
                  "width": "100%",
                },
              ]
            }
          >
            Username or E-mail address:
          </Text>
          <TextInput
            field="Username or E-mail address:"
            name="username"
            onBlur={[Function]}
            onChangeText={[Function]}
            placeholder=""
            returnKeyLabel="next"
            returnKeyType="next"
            style={
              Object {
                "borderBottomColor": "#d8d8d8",
                "borderBottomWidth": 1,
                "fontSize": 17,
                "height": 40,
              }
            }
            testID="username-input"
            value=""
          />
        </View>
        <View>
          <Text
            style={
              Array [
                Object {
                  "fontFamily": "SourceSansPro_400Regular",
                },
                Object {
                  "fontSize": 15,
                  "lineHeight": 19,
                  "paddingBottom": 5,
                  "paddingTop": 10,
                  "width": "100%",
                },
              ]
            }
          >
            Password:
          </Text>
          <TextInput
            field="Password:"
            name="password"
            onBlur={[Function]}
            onChangeText={[Function]}
            placeholder=""
            returnKeyLabel="next"
            returnKeyType="next"
            secureTextEntry={true}
            style={
              Object {
                "borderBottomColor": "#d8d8d8",
                "borderBottomWidth": 1,
                "fontSize": 17,
                "height": 40,
              }
            }
            testID="password-input"
            value=""
          />
        </View>
        <Text
          onPress={[Function]}
          style={
            Array [
              Object {
                "fontFamily": "SourceSansPro_400Regular",
              },
              Array [
                Object {
                  "textDecorationLine": "underline",
                },
                Object {
                  "fontSize": 16,
                  "marginTop": 60,
                },
              ],
            ]
          }
        >
          Forgot Password?
        </Text>
      </View>
    </RCTScrollView>
  </View>
  <View
    style={
      Array [
        Object {
          "overflow": "hidden",
        },
        Object {
          "borderRadius": 2,
        },
        undefined,
        false,
      ]
    }
    testID="RNE_BUTTON_WRAPPER"
  >
    <View
      accessibilityRole="button"
      accessibilityState={
        Object {
          "busy": false,
          "disabled": false,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#2d40cc",
            "borderColor": "#2089dc",
            "borderRadius": 0,
            "borderWidth": 0,
            "flexDirection": "row",
            "fontSize": 16,
            "height": 70,
            "justifyContent": "center",
            "padding": 8,
            "paddingHorizontal": 10,
            "width": "100%",
          }
        }
      >
        <Text
          style={
            Object {
              "color": "white",
              "fontSize": 18,
              "paddingVertical": 1,
              "textAlign": "center",
            }
          }
        >
          Login
        </Text>
      </View>
    </View>
  </View>
</View>
`;
